tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -741,6 +755,691 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
 	}
 }
 
+/*
+ * Check if the join fromList belongs to a INNER JOIN / LEFT JOIN
+ * on two tables. A INNER JOIN / LEFT JOIN on two tables has one JoinExpr
+ * entry in fromList.
+ */
+static bool
+is_inner_or_left_join_two_tables(List *fromList)
+{
+	if (!IsA(list_nth(fromList, 0), JoinExpr) || (list_length(fromList) != 1))
+		return false;
+
+	return true;
+}
+
+/*
+ * Check if the join fromList belongs to an implicit join (e.g.,
+ * SELECT * FROM metric m1, metric_name m2 WHERE m1.id=m2.id). This type
+ * of join has two RANGETBLREF entries in the fromList.
+ */
+static bool
+is_implicit_join_two_tables(List *fromList)
+{
+	if (list_length(fromList) != 2)
+		return false;
+
+	/* At the moment, the code is called only in the reference table join
+	 * pushdown path. Earlier checks ensure that both parameters are
+	 * RangeTblRefs. Recheck the condition here and handle the case properly.
+	 */
+	if (!IsA(list_nth(fromList, 0), RangeTblRef))
+		return false;
+
+	if (!IsA(list_nth(fromList, 1), RangeTblRef))
+		return false;
+
+	return true;
+}
+
+/*
+ * Check if the query performs a join between a hypertble (outer) and a reference
+ * table (inner) and the join type is a LEFT JOIN, an INNER JOIN, or an implicit
+ * join.
+ */
+static bool
+is_safe_to_pushdown_reftable_join(PlannerInfo *root, List *join_reference_tables,
+								  RangeTblEntry *innertableref, JoinType jointype)
+{
+	Query *query = root->parse;
+	List *fromList;
+
+	Assert(root != NULL);
+	Assert(innertableref != NULL);
+
+	/*
+	 * We support pushing down of INNER and LEFT joins only.
+	 *
+	 * Constructing queries representing partitioned OUTER, SEMI and ANTI
```

Comment:
```suggestion
	 * Constructing queries representing partitioned FULL, SEMI and ANTI
```

Did you mean FULL? The LEFT and RIGHT are both a variants of outer.

Reply:
Thanks, you are correct, the comment has been updated.